File: backend/internal/pricing/promotion/entity/target_type.go
package entity

import "fmt"

// TargetType represents the type of object being promoted.
// Only these three types are supported for promotion.
type TargetType string

const (
	// TargetTypeForSale promotes a fixed-price sale.
	TargetTypeForSale TargetType = "for_sale"

	// TargetTypeAuction promotes an auction.
	TargetTypeAuction TargetType = "auction"

	// TargetTypeExternalProduct promotes an external product (URL).
	TargetTypeExternalProduct TargetType = "external_product"
)

// IsValid returns true if the target type is valid.
func (t TargetType) IsValid() bool {
	switch t {
	case TargetTypeForSale, TargetTypeAuction, TargetTypeExternalProduct:
		return true
	default:
		return false
	}
}

// RequiresTargetID returns true if this target type needs a target_id reference.
// External products now point at external_products.id so they can be hydrated
// from the reviewable product/media model.
func (t TargetType) RequiresTargetID() bool {
	return t == TargetTypeForSale || t == TargetTypeAuction || t == TargetTypeExternalProduct
}

// IsPublicPromotable returns true when this target type can be shown on public discovery surfaces.
func (t TargetType) IsPublicPromotable() bool {
	return t == TargetTypeForSale || t == TargetTypeAuction || t == TargetTypeExternalProduct
}

// String returns the string representation of the target type.
func (t TargetType) String() string {
	return string(t)
}

// InvalidTargetTypeError is returned when an invalid target type is provided.
type InvalidTargetTypeError struct {
	TargetType TargetType
}

func (e *InvalidTargetTypeError) Error() string {
	return fmt.Sprintf("invalid target type: %s", e.TargetType)
}

File: apps/mobile/lib/domains/commerce/pricing/promotion/domain/entities/target_type.dart
/// Target type enum for promotion.
///
/// Defines what can be promoted:
/// - for_sale: A fixed-price sale
/// - auction: An auction
/// - externalProduct: An external product (URL-based)
enum TargetType {
  /// A fixed-price sale can be promoted
  forSale('for_sale'),

  /// An auction can be promoted
  auction('auction'),

  /// An external product (with embedded URL, title, media) can be promoted
  externalProduct('external_product');

  const TargetType(this.value);

  /// The API value for this enum
  final String value;

  /// Parses target type from string
  static TargetType fromString(String value) {
    return TargetType.values.firstWhere(
      (type) => type.value == value,
      orElse: () => TargetType.forSale,
    );
  }
}

/// Extension for TargetType utility methods
extension TargetTypeX on TargetType {
  /// Whether this target type requires a target_id reference
  bool get requiresTargetId =>
      this == TargetType.forSale || this == TargetType.auction;

  /// Whether this target type requires external fields (URL, title, media)
  bool get requiresExternalFields => this == TargetType.externalProduct;
}
